File: Codes/076_minWindow/76.最小覆盖子串.py
'''
Description: 
Autor: HTmonster
Date: 2022-02-18 11:21:39
'''
#
# @lc app=leetcode.cn id=76 lang=python3
#
# [76] 最小覆盖子串
#

# @lc code=start
class Solution:
    def minWindow(self, s: str, t: str) -> str:

        charIndex={}   # 字符索引
        need=[]
        i=0
        for c in t:
            if c not in charIndex.keys():
                charIndex[c]=i
                i+=1
                need.append(1)
            else:
                need[charIndex[c]]+=1

        # 结果开始点 和结尾
        res_left,res_right=0,float("inf")

        left,right=0,0 # 左右指针  window=[left,right]
        containCnt=[0 for _ in range(len(charIndex.keys()))]  # 包含的字符计数

        def check():
            nonlocal need,check
            for i in range(len(need)):
                if need[i] > containCnt[i]:
                    return False
            return True

        while right <len(s):
            #1. 右边先行 直到包含所有字符串后停止
            while not check() and right < len(s):
                if s[right] in t: containCnt[charIndex[s[right]]]+=1
                right+=1
            
            # print([left, right,s[left:right]],containCnt)

            #2. 左边收缩 尝试最小的范围
            while check() and left< len(s):
                if s[left] in t: containCnt[charIndex[s[left]]]-=1
                #更新值
                if right-left<res_right-res_left:
                    res_left,res_right =left,right

                left +=1
                # print("*", [left, right,s[left:right]],containCnt)
        
        if res_right==float("inf"):
            return ""
        else:
            return s[res_left:res_right]
# @lc code=end

File: Codes/076_minWindow/76.最小覆盖子串.go
/*
 * @Description:
 * @Autor: HTmonster
 * @Date: 2022-02-18 15:09:05
 */
/*
 * @lc app=leetcode.cn id=76 lang=golang
 *
 * [76] 最小覆盖子串
 */

// @lc code=start
func minWindow(s string, t string) string {

	// 先建立一个字符的索引，以及需求的数组
	tIndex := make(map[byte]int)
	need := make([]int, 0) // 目标需要的计数
	cnt := make([]int, 0)  // 已经包含的计数

	i := 0
	for j := 0; j < len(t); j++ {
		c := t[j]
		//已经包含了 例如 "aa"
		if _, ok := tIndex[c]; ok {
			need[tIndex[c]]++
		} else {
			tIndex[c] = i
			need = append(need, 1)
			cnt = append(cnt, 0)
			i++
		}
	}

	fmt.Println(tIndex, need)

	// 检查函数 检查是否满足要求
	var check func() bool
	check = func() bool {
		for i := range need {
			// 其中一项还没达到要求
			if need[i] > cnt[i] {
				return false
			}
		}
		return true
	}

	ansLeft, ansRight := 0, math.MaxInt32 // 答案的范围
	left, right := 0, 0                   // 窗口

	for right < len(s) {
		//1. 窗口扩大
		for !check() && right < len(s) {
			// 包含了一个字符
			if index, ok := tIndex[s[right]]; ok {
				cnt[index]++
			}
			right++
		}
		// fmt.Println(left, right, s[left:right], cnt)

		//2.窗口缩小
		for check() && left < right {
			//更新值
			if right-left < ansRight-ansLeft {
				ansLeft = left
				ansRight = right
			}

			// 去除一个字符
			if index, ok := tIndex[s[left]]; ok {
				cnt[index]--
			}
			left++
			// fmt.Println("*", left, right, s[left:right], cnt)
		}
	}

	if ansRight == math.MaxInt32 {
		return ""
	} else {
		return s[ansLeft:ansRight]
	}
}

// @lc code=end

